[docker_network] Adding `scope` and `attachable` flags

DBendit opened the pull request:
##### SUMMARY
Incorporating the abandoned work from PRs #35288 and #45552. Also adding in
the version checking from `docker_container.py`, which should be abstracted
out to `docker_common.py`.

Part of the work for https://github.com/ansible/community/issues/408

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_network

##### ADDITIONAL INFORMATION
Integration tests passing with `ansible-test integration -v docker_network --docker ubuntu1604`

DBendit committed:
[docker_network] Adding `scope` and `attachable` flags

Incorporating the abandoned work from PRs #35288 and #45552. Also adding in
the version checking from `docker_container.py`, which should be abstracted
out to `docker_common.py`.

DBendit commented on test/integration/targets/docker_network/tasks/tests/options.yml:
```diff
@@ -92,3 +92,93 @@
+  - name: scope (change)
+    docker_network:
+      name: "{{ nname_1 }}"
+      driver: overlay
```
If there's a better way to do this, please let me know. I don't think docker includes any drivers that support multiple modes (or at least, I wasn't able to figure out how to get any of them to switch in a way that was reflected in the output - swarm and global both report as swarm for some reason, and overlay networks can't downgrade to local).

felixfontein replied:
I think you should mention that as a comment in the test, so it's clear that this is not an oversight, but an conscious (unfortunate) decision you had to make.

DBendit commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -294,6 +316,7 @@ def get_ip_version(cidr):
 
 
 def get_driver_options(driver_options):
+    # TODO: Move this and the same from docker_prune.py to docker_common.py
```
Created https://github.com/ansible/ansible/issues/49563 to track

DBendit commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -310,6 +333,59 @@ def get_driver_options(driver_options):
 class DockerNetworkManager(object):
 
+    def _get_minimal_versions(self):
+        # TODO: Move this and the same from docker_container.py to docker_common.py
```
Created https://github.com/ansible/ansible/issues/49564 to track

felixfontein commented on test/integration/targets/docker_network/tasks/tests/options.yml:
```diff
@@ -92,3 +92,93 @@
+      - attachable_3 is changed
+
+  # Requirements for attachable parameter
+  when: docker_py_version is version('2.0.0', '>=') and docker_api_version is version('1.25', '>=')
```
How about making the test also work for earlier docker-py versions (similar to the `docker_container` tests), i.e. that the minimum version check is also tested?

DBendit commented:
Added comment to the test to indicate why it doesn't strictly do what it says it does (sigh).

ready_for_review

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -310,6 +333,59 @@ def get_driver_options(driver_options):
+            self.option_minimal_versions[option] = dict()
+        self.option_minimal_versions.update(dict(
+            scope=dict(docker_py_version='2.6.0', docker_api_version='1.29'),
+            attachable=dict(docker_py_version='2.0.0', docker_api_version='1.25'),
```
I couldn't find this in the [1.25 API documentation](https://docs.docker.com/engine/api/v1.25/#operation/NetworkCreate), but only in the [1.26 API documentation](https://docs.docker.com/engine/api/v1.26/#operation/NetworkCreate). Interestingly, the docker-py source code claims it has been supported since 1.24...

DBendit replied:
https://docs.docker.com/engine/reference/commandline/network_create/ claims it's available on 1.25+. I'll set the minimum to 1.26 just to be safe.

felixfontein replied:
Hmm, that's pretty inconsistent. So now we got three different choices... :) I guess it's fine to leave it with 1.25.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -310,6 +333,59 @@ def get_driver_options(driver_options):
+        for option, data in self.client.module.argument_spec.items():
+            self.option_minimal_versions[option] = dict()
+        self.option_minimal_versions.update(dict(
+            scope=dict(docker_py_version='2.6.0', docker_api_version='1.29'),
```
This one I cannot find in any docker API documentation (I checked up to 1.38). According to the docker-py code, it's supported from 1.30 on. Where did you got the information about the minimal API levels for these two options?

DBendit replied:
https://docs.docker.com/engine/api/version-history/#v129-api-changes

"GET /networks/ now supports a scope filter to filter networks based on the network mode (swarm, global, or local)."

Edit: My thought was that if it was available to be retrieved, it must be available to be set...right?

DBendit replied:
You know what, I'll also increase this one to 1.30, just in case the docker-py folks know something we don't.

felixfontein replied:
The 1.29 API change might only refer to allowing to *filter* by scope, though I'd be surprised that the API only allows to filter when you can't *create* a network with a given scope.

It looks to me more and more like the docker API docs are kind of a mess ;-) Anyway, going to 1.30 is probably safe. If someone reports it also works with 1.29, we can still change it back.

felixfontein commented:
The current CI failure is unrelated (pip fails installing requirements).

felixfontein commented:
BTW, instead of making trivial changed, you can also close and reopen the PR to trigger a new CI run.

DBendit commented:
Good to know on the closing/opening trick to get the tests to re-run, especially since trailing whitespace causes the above issues.

I've updated the minimum versions and added the support for old version testing.

ready_for_review

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -132,6 +132,26 @@
+  scope:
+    version_added: 2.8
+    description:
+      - Specify the network's scope (C(local), C(global), or C(swarm)).
```
Why repeat the values here, when they are already listed left to this (in the generated HTML docs)?

DBendit replied:
I took the verbiage straight out of docker-py's docs. I can get rid of the values.

felixfontein replied:
They don't have a place for possible values like we have, so for them it makes sense to list the options inline. I think here it's fine to just have them once :) (Also reduces the chance of forgetting to update one list in the future...)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -345,6 +421,8 @@ def __init__(self, client):
                 self.diff_result['before'], self.diff_result['after'] = self.diff_tracker.get_before_after()
             self.results['diff'] = self.diff_result
 
+        self._get_minimal_versions()
```
This must be moved up before the module logic (line ~400). Otherwise, the module will break with a stacktrace (or other error) if a non-supported option is used.

DBendit replied:
Good catch.

felixfontein reviewed (approved):
shipit

I've ran the integration tests with various docker-py versions (1.10.6, 2.0.0, 2.5.0 and 3.6.0), everything looks good!

DBendit commented:
@felixfontein Thanks for the reviews!
@gundalow Thanks for the merge!

felixfontein commented:
@DBendit thanks for the PR!
@gundalow thanks for merging!

lightster commented:
Thank you @felixfontein!
